source/extensions/filters/http/on_demand/on_demand_update.cc

```diff
@@ -28,6 +64,57 @@ Http::FilterHeadersStatus OnDemandRouteUpdate::decodeHeaders(Http::RequestHeader
   return filter_iteration_state_;
 }
 
+// Passed route pointer here is not null.
+void OnDemandRouteUpdate::handleOnDemandCDS(const Router::RouteConstSharedPtr& route) {
+  auto config = getConfig(route);
+  if (config == nullptr) {
+    filter_iteration_state_ = Http::FilterHeadersStatus::Continue;
+    return;
+  }
+  auto odcds = config->odcds();
+  if (odcds == nullptr) {
+    // No ODCDS? It means that on-demand discovery is disabled.
+    filter_iteration_state_ = Http::FilterHeadersStatus::Continue;
+    return;
+  }
+  if (callbacks_->clusterInfo() != nullptr) {
+    // Cluster already exists, so nothing to do here.
+    filter_iteration_state_ = Http::FilterHeadersStatus::Continue;
+    return;
+  }
+  auto entry = route->routeEntry();
+  if (entry == nullptr) {
+    // No entry? Nothing we can do here.
+    filter_iteration_state_ = Http::FilterHeadersStatus::Continue;
+    return;
+  }
+  auto cluster_name = entry->clusterName();
+  if (cluster_name.empty()) {
+    // Empty cluster name may be a result of a missing HTTP header
+    // used for getting the cluster name. Nothing we can do here.
+    filter_iteration_state_ = Http::FilterHeadersStatus::Continue;
+    return;
+  }
+  filter_iteration_state_ = Http::FilterHeadersStatus::StopIteration;
+  auto callback = std::make_unique<Upstream::ClusterDiscoveryCallback>(
+      [this](Upstream::ClusterDiscoveryStatus cluster_status) {
+        onClusterDiscoveryCompletion(cluster_status);
```

Comment:
Could there be a case where the filter is destroyed before the callback is invoked?
What I'm not certain about is whether we'll get an invocation of `onClusterDiscoveryCompletion` on a destroyed filter.

Reply:
If the filter is destroyed, then `cluster_discovery_handle_` is going to be destroyed too, which will unregister and destroy the callback.